### PR TITLE
PIM-5067 : add a custom JUnit formatter

### PR DESCRIPTION
**Pull request by jmleroux**
| Q                 | A
| ----------------- | ---
| Specs             |no
| Behats            |yes
| Blue CI           |yes
| Changelog updated |yes
| Review and 2 GTM  |
| Micro Demo (PO)   |
| Migration script  |no
| Tech Doc          |no

### CODE REVIEW

**Comment by jjanvier** [comment]
gtm

File: features/Context/Formatter/JUnitFormatter.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    /**
+     * @param FeatureNode $feature
+     *
+     * @return string
+     */
+    private function formatClassname(FeatureNode $feature)
+    {
+        $filepath = htmlspecialchars($feature->getFile());
+        $classname = substr($filepath, strpos($filepath,'features/') + strlen('features/'));
+        return str_replace('/', '__', $classname);
```
Comment (ClementGautier) [inline features/Context/Formatter/JUnitFormatter.php]: missing blank line before return statement

File: features/Context/Formatter/JUnitFormatter.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            ? ', Ex #' . ($event->getIteration() + 1)
+            : '';
+        $caseStats = sprintf('classname="%s" name="%s" time="%F" assertions="%d"',
+            htmlspecialchars($className),
+            htmlspecialchars($name),
+            $time,
+            $this->scenarioStepsCount
+        );
+
+        $xml  = "    <testcase $caseStats>\n";
```
Comment (ClementGautier) [inline features/Context/Formatter/JUnitFormatter.php]: you could build the XML using SimpleXMLElement for exemple

File: features/Context/Formatter/JUnitFormatter.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+use Behat\Gherkin\Node\FeatureNode;
+use Behat\Gherkin\Node\ScenarioNode;
+
+class JUnitFormatter extends BehatJUnitFormatter
+{
+    protected function printTestCase(ScenarioNode $scenario, $time, EventInterface $event)
+    {
+        $className = $this->formatClassname($scenario->getFeature());
+        $name = $scenario->getTitle();
+        $name .= $event instanceof OutlineExampleEvent
```
Comment (ClementGautier) [inline features/Context/Formatter/JUnitFormatter.php]:
```php
if ($event instanceof OutlineExampleEvent) {
    $name = sprintf('%s, Ex #%d', $name, $event->getIteration() + );
}
```

is more readable for me

File: features/Context/Formatter/JUnitFormatter.php (hunk 2 above)
Comment (ClementGautier) [inline features/Context/Formatter/JUnitFormatter.php]: using SimpleXMLElement would remove a lot of sprintf in this file

File: features/Context/Formatter/JUnitFormatter.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                $elemAttributes
+            );
+            $exception = str_replace(array('<![CDATA[', ']]>'), '', (string) $exception);
+            $xml .= sprintf(
+                "<![CDATA[\n%s\n]]></%s>\n",
+                $exception,
+                $elemType
+            );
+        }
+        $this->exceptions = array();
```
Comment (ClementGautier) [inline features/Context/Formatter/JUnitFormatter.php]: array => []

**Comment by ClementGautier** [comment]
I'll do an other review after the rework with SimpleXMLElement if you do it before giving my GTM

**Comment by jmleroux** [comment]
@ClementGautier , i will not use Simple XML to be just like the extended behat formatter.

File: features/Context/Formatter/JUnitFormatter.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            $this->scenarioStepsCount
+        );
+
+        $xml = "    <testcase $caseStats>\n";
+
+        foreach ($this->exceptions as $exception) {
+            $error          = $this->exceptionToString($exception);
+            $elemType       = $this->getElementType($event->getResult());
+            $elemAttributes = '';
+            if ($elemType !== 'skipped') {
```
Comment (sumbobyboys) [inline features/Context/Formatter/JUnitFormatter.php]: yoda code

**Comment by sumbobyboys** [comment]
GTM

**Pull request merged**